main.cpp

```diff
@@ -131,41 +202,190 @@ std::pair<JSONObject, size_t> parse(std::string_view json) {
         }
         return {JSONObject{std::move(res)}, i};
     } else if (json[0] == '{') {
+        //for dict
         std::unordered_map<std::string, JSONObject> res;
         size_t i;
         for (i = 1; i < json.size();) {
             if (json[i] == '}') {
                 i += 1;
                 break;
             }
-            auto [keyobj, keyeaten] = parse(json.substr(i));
-            if (keyeaten == 0) {
-                i = 0;
-                break;
-            }
-            i += keyeaten;
-            if (!std::holds_alternative<std::string>(keyobj.inner)) {
-                i = 0;
-                break;
-            }
-            if (json[i] == ':') {
-                i += 1;
+            std::string key;
+            {//for key
+                std::regex elem_re(R"([ \n\r\t\v\f\0]*?[\'\"]{0,1}[a-zA-Z]+[\'\"]{0,1}[ \n\r\t\v\f\0]*?:)");
+                std::cmatch match;
+                std::regex_search(json.data() + i, json.data() + json.size(), match, elem_re);
+                key = match.str();
+                i += key.size();
+                key = key.substr(0, key.size() - 1);
+                int beg = 0;
+                while(std::string(" \n\r\t\v\f\0").find(key.at(beg)) != std::string::npos){
+                    beg++;
+                }
+                int end = key.size() - 1;
+                while(std::string(" \n\r\t\v\f\0").find(key.at(end)) != std::string::npos){
+                    end--;
+                }
+
+                key = key.substr(beg, end - beg + 1);
             }
-            std::string key = std::move(std::get<std::string>(keyobj.inner));
             auto [valobj, valeaten] = parse(json.substr(i));
             if (valeaten == 0) {
                 i = 0;
                 break;
             }
             i += valeaten;
             res.try_emplace(std::move(key), std::move(valobj));
+            //solve issue : can't allow any space behind value
+            while(std::string(" \n\r\t\v\f\0").find(json[i])!=std::string::npos){
+                i++;
+            }
```

Comment:
Did you realize that i may run out-of-bound in json and you use unsafe []? It doesn't run out of bound just thank to the fact that json is string which always end in `"\0"` in libstdc++.
```suggestion
i = json.find_first_not_of(std::string_view(" \n\r\t\v\f\0"), i);
```

Reply:
yes, realized a serious issue... `json.find_first_not_of(" \n\r\t\v\f\0", i)` is eqv to `json.find_first_not_of(" \n\r\t\v\f", i)`, must use `json.find_first_not_of(" \n\r\t\v\f\0"sv, i)` to ensure zero is trimed too.